File: backend/src/main/scala/models/Recipe.scala
package io.dmtri.minecraft
package models

sealed trait Recipe {
  def id: Int
  def recipeType: String
  def recipeCategory: Option[String]
  def recipeGroup: Option[String]
  def resultItem: Item
  def resultItemAmount: Int
}

object Recipe {
  case class CraftingShaped(
      id: Int,
      recipeCategory: Option[String],
      recipeGroup: Option[String],
      ingredients: Map[String, Seq[Item]],
      resultItem: Item,
      resultItemAmount: Int,
      craftPattern: String,
      likeStatus: LikeStatus = LikeStatus.NoStatus
  ) extends Recipe {
    override def recipeType: String =
      if (
        craftPattern.count(_ == '\n') >= 3 || craftPattern
          .split('\n')
          .exists(_.length >= 3)
      ) "shaped-3"
      else "shaped-2"
  }

  case class CraftingShapeless(
      id: Int,
      recipeCategory: Option[String],
      recipeGroup: Option[String],
      shapelessIngredients: Seq[Seq[Item]],
      resultItem: Item,
      resultItemAmount: Int,
      likeStatus: LikeStatus = LikeStatus.NoStatus
  ) extends Recipe {
    override def recipeType: String = "shapeless"
  }

  case class Smelting(
      id: Int,
      recipeCategory: Option[String],
      recipeGroup: Option[String],
      sourceItem: Item,
      resultItem: Item,
      resultItemAmount: Int,
      smeltTime: Int,
      smeltType: String,
      likeStatus: LikeStatus = LikeStatus.NoStatus
  ) extends Recipe {
    override def recipeType: String = "smelting"
  }

  case class StoneCutter(
      id: Int,
      recipeCategory: Option[String],
      recipeGroup: Option[String],
      resultItem: Item,
      resultItemAmount: Int,
      sourceItem: Item,
      likeStatus: LikeStatus = LikeStatus.NoStatus
  ) extends Recipe {
    override def recipeType: String = "stonecutter"
  }

  case class Other(
      id: Int,
      recipeCategory: Option[String],
      recipeGroup: Option[String],
      resultItem: Item,
      resultItemAmount: Int,
      ingredients: Map[String, Seq[Item]],
      likeStatus: LikeStatus = LikeStatus.NoStatus
  ) extends Recipe {
    override def recipeType: String = "unknown"
  }
}

File: frontend/src/models/Recipe.ts
interface IngredientsMap {
  [key: string]: Item[]
}

export type LikeStatus = "like" | "dislike" | "no_status"

export type Recipe = {
  CraftingShaped?: CraftingShaped
  CraftingShapeless?: CraftingShapeless
  Smelting?: Smelting
  Other?: Other
}

export type BaseRecipe = {
  id: number
  resultItem: Item
  resultItemAmount: number
  recipeCategory: string
  recipeGroup: string
  likeStatus: LikeStatus
}

export type CraftingShaped = {
  ingredients: IngredientsMap
  craftPattern: string
} & BaseRecipe

export type CraftingShapeless = {
  shapelessIngredients: Item[][]
  resultItem: Item
  resultItemAmount: number
} & BaseRecipe

export type Smelting = {
  sourceItem: Item
  smeltTime: number
  smeltType: "smelting" | "blasting"
} & BaseRecipe

export type Other = {
  ingredients: IngredientsMap
} & BaseRecipe

export function extractRecipe(recipe: Recipe): BaseRecipe {
  const baseRecipe =
    recipe.CraftingShaped ||
    recipe.CraftingShapeless ||
    recipe.Smelting ||
    recipe.Other

  if (baseRecipe === undefined)
    throw new Error("Empty recipe, this should not happen")

  return baseRecipe
}
